src/notification.cpp

```diff
@@ -52,6 +58,145 @@ NotificationManagerProxy *notificationManager()
     return notificationManagerProxyInstance();
 }
 
+QString encodeDBusCall(const QString &service, const QString &path, const QString &iface, const QString &method, const QVariantList &arguments)
+{
+    const QString space(QStringLiteral(" "));
+
+    QString s = service % space % path % space % iface % space % method;
+
+    if (!arguments.isEmpty()) {
+        QStringList args;
+        int argsLength;
+
+        foreach (const QVariant &arg, arguments) {
+            // Serialize the QVariant into a Base64 encoded byte stream
+            QByteArray buffer;
+            QDataStream stream(&buffer, QIODevice::WriteOnly);
+            stream << arg;
+            args.append(space + buffer.toBase64());
+            argsLength += args.last().length();
+        }
+
+        s.reserve(s.length() + argsLength);
+        foreach (const QString &arg, args) {
+            s.append(arg);
+        }
+    }
+
+    return s;
+}
+
+QStringList encodeActions(const QHash<QString, QString> &actions)
+{
+    QStringList rv;
+
+    // Actions are encoded as a sequence of name followed by displayName
+    QHash<QString, QString>::const_iterator it = actions.constBegin(), end = actions.constEnd();
+    for ( ; it != end; ++it) {
+        rv.append(it.key());
+        rv.append(it.value());
+    }
+
+    return rv;
+}
+
+QHash<QString, QString> decodeActions(const QStringList &actions)
+{
+    QHash<QString, QString> rv;
+
+    QStringList::const_iterator it = actions.constBegin(), end = actions.constEnd();
+    for ( ; it != end; ++it) {
+        const QString &key(*it);
+        if (++it == end) {
```

Comment:
If there's an uneven amount of items there, that would be an error, right? Maybe do a warning log call here or something instead of just silently dropping the last item.